tests/test_events.py

```diff
@@ -50,6 +52,30 @@ def test_wait_for(client_or_bot: disnake.Client, event) -> None:
     coro.close()  # close coroutine to avoid warning
 
 
+def _test_typing_wait_for(client: disnake.Client, bot: commands.Bot) -> None:
+    expected_type = Coroutine[Any, Any, disnake.Guild]
+
+    # valid enum event
+    _ = assert_type(client.wait_for(Event.guild_join), expected_type)
+    _ = assert_type(client.wait_for(Event.guild_join, check=lambda g: True), expected_type)
+
+    # valid str event
+    _ = assert_type(client.wait_for("guild_join"), expected_type)
+    _ = assert_type(client.wait_for("guild_join", check=lambda g: True), expected_type)
+
+    # invalid check type
+    _ = client.wait_for(Event.guild_join, check=lambda: True)  # type: ignore
+    # n.b. this one isn't ideal, but there's no way to prevent type-checkers from using the fallback in this case
+    _ = assert_type(client.wait_for("guild_join", check=lambda: True), Coroutine[Any, Any, Any])
```

Comment:
Technically, an overload like
```py
@overload
async def wait_for(
    self,
    event: Literal["guild_join", "ready", ...] | Event,
    *,
    check: Callable[..., Any],
    timeout: Optional[float] = None,
) -> Never: ...
```
Could work, but the `Literal` would be huge, and it's non-obvious what that signature means/why it got matched

Reply:
oh, interesting idea

And yeah, communicating this clearly to the user is tough. I could see `InvalidCheckType = Never` (or w/e) maybe being an option, which would look something like this, but it's not pretty:
![image](https://github.com/user-attachments/assets/9e692c3e-7ec6-4fa4-a969-d0f451966c8e)

fwiw, I think ultimately the approach in this PR is somewhat flawed, but it's the best we can do for now, sadly. In an ideal world ~~(3.0?)~~, every event would be a dataclass/namedtuple, and event names + overloads would be a thing of the past.